### PR TITLE
Update workshop for StateFun 3.0

### PR DESCRIPTION
**Pull request by sjwiesman**
This updates the workshop to StateFun 3.0 and simplifies the deployment process so functions are run directly in the user's IDE. Please see the README for details on how to build and execute. 

This is the final solution, we will not merge this into master but its own branch.

### CODE REVIEW

**Comment by alpinegizmo** [comment]
The image failed to build for me. I'm not clear on how these two PRs are meant to fit together, and how to work with this setup ATM.

**Comment by sjwiesman** [comment]
The simulator only exists on the other branch. You need to checkout out simulator, build and tag the image using the command, and then checkout this branch to run the workshop. Sorry its a little confusing. After the simulator is merged in I will publish it to a registry and have the docker-compose reference that so participants will not need to build anything locally.

File: docker-compose.yml [hunk 1]
```diff
@@ -1,78 +1,94 @@
...
-      - "6121"
-      - "6122"
+      PORTS: 8000
+
+  ###############################################################
+  #    Simple Kafka JSON producer to simulate ingress events
+  ###############################################################
+
+  simulator:
+    image: statefun-workshop-simulator:1.0.0
```
Comment (sjwiesman) [inline docker-compose.yml]: Once #11 is published I will update this so the image does not need to be built locally

File: workshop/__main__.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    This function tracks the total number of reported fraudulent transactions made against an account
+    on a rolling 30 minute period. It supports three message types:
+
+    1) ConfirmFraud: When a customer reports a fraudulent transaction, the function will
+    receive this message. It will increment its internal count and set a 30 minute expiration timer.
+
+    2) ExpireFraud: After 30 minutes, the function will receive an expiration message. At this time it
+    will decrement its internal count.
+
+    3) QueryFraud: After 30 minutes, the function will receive an expiration message. At this tme
```
Comment (alpinegizmo) [inline workshop/__main__.py]:
```suggestion
    3) QueryFraud: After 30 minutes, the function will receive an expiration message. At this time
```

**Review by alpinegizmo — COMMENTED** [review]
I have the impression that the final (Architecture) section of the README hasn't been updated.

I feel like all of the code that users will study needs more comments: models.py, __main__.py, and the tests. I'm currently looking at this without a thorough understanding of how the new python SDK works -- and I'm confused.

**Comment by sjwiesman** [comment]
I've added many more substantial comments to the relevant files. I will update the README tomorrow.

File: workshop/__main__.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    req = await request.read()
+    res = await handler.handle_async(req)
+    return web.Response(body=res, content_type='application/octet-stream')
+
+
+app = web.Application()
+app.add_routes([web.post('/statefun', handle)])
+
+if __name__ == '__main__':
+    web.run_app(app, port=8000)
```
Comment (afedulov) [inline workshop/__main__.py:249]: This collides with the `flink-training-slides` port. Let's maybe pick something else?

**Comment by sjwiesman** [comment]
I've merged and published the simulator. This PR now depends on a published image so you do not need to build anything locally first.

File: images/statefun-workshop.svg [hunk 4]
```diff
@@ -1,320 +1,378 @@
...
+       y="-423.22107"
+       x="35.831715"
+       transform="scale(1.0001444,-0.99985561)"
+       style="font-variant:normal;font-weight:normal;font-size:13.998px;font-family:Arial;-inkscape-font-specification:ArialMT;writing-mode:lr-tb;fill:#ffffff;fill-opacity:1;fill-rule:nonzero;stroke:none;stroke-width:0.239965"
+       id="text184"><tspan
+         style="stroke-width:0.239965"
+         x="35.831715 46.257408 54.275452 62.293495 66.308113 69.510849 74.323357 86.358818 94.376854 102.39491 106.40952 115.22545 120.03795 128.05598 136.07404"
+         y="-423.22107"
+         sodipodi:role="line"
+         id="tspan182">Confirmed Fraud</tspan></text><image
```
Comment (afedulov) [inline images/statefun-workshop.svg:301]: I assume one of those input queues is supposed to have another name (transactions?)

File: workshop/models.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+# JSON serialized representation and python
+# classes. `simple_type` defines the serializer,
+# deserializer, and type name for the corresponding
+# python classes.
+#####################################################
+
+
+class Transaction:
+    """
+    A Transaction event which triggers the pipeline and is tested authenticity.
```
Comment (afedulov) [inline workshop/models.py:33]: `is tested FOR authenticity`?

File: workshop/__main__.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+from aiohttp import web
+from statefun import *
+
+THRESHOLD = 1
+
+logger = logging.getLogger('workshop')
+
+#####################################################
+# `functions` is the `RequestReplyHandler` to which
+# all user defined functions are be bound. It
```
Comment (afedulov) [inline workshop/__main__.py:34]: `are being` or just `are`?

File: workshop/__main__.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+    This function tracks the total number of reported fraudulent transactions made against an account
+    on a rolling 30 minute period. It supports three message types:
+
+    1) ConfirmFraud: When a customer reports a fraudulent transaction, the function will
+    receive this message. It will increment its internal count and set a 30 minute expiration timer.
+
+    2) 'expire': After 30 minutes, the function will receive an expiration message. At this time it
+    will decrement its internal count.
+
+    3) 'query': After 30 minutes, the function will receive an expiration message. At this time
```
Comment (afedulov) [inline workshop/__main__.py:70]: Copy-paste from "expire"?

File: workshop/__main__.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+@functions.bind(
+    'com.ververica.fn/counter',
+    [ValueSpec(name='fraud_count', type=IntType)])
+async def fraud_count(ctx: Context, message: Message):
+    """
+    This function tracks the total number of reported fraudulent transactions made against an account
+    on a rolling 30 minute period. It supports three message types:
+
+    1) ConfirmFraud: When a customer reports a fraudulent transaction, the function will
+    receive this message. It will increment its internal count and set a 30 minute expiration timer.
```
Comment (afedulov) [inline workshop/__main__.py:65]: Maybe rephrase to something like `and will schedule an 'expire' message`. Otherwise, it is hard to understand why you need both 1) and 2).

File: workshop/__main__.py [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+    else:
+        logger.warning(f"Unknown {message.value_typename()}")
+
+
+def merchant_scorer(client=third_party_api_client):
+    """
+    This function queries a 3rd party API to retrieve a trustworthiness
+    score for the merchant. The score will be stored in state, as a sort of
+    cache, to reduce expensive API calls. The state is set to expire 1 hour
+    after write to prevent the use of state data.
```
Comment (afedulov) [inline workshop/__main__.py:106]:
```suggestion
    after write to prevent the use of stale data.
```

**Review by afedulov — APPROVED** [review]
I really like this example/task, great job. I marked some minor typos. Do you already have a formulation of the actual task? What will the attendees be supposed to do?

File: images/statefun-workshop.svg [hunk 10]
```diff
@@ -1,320 +1,378 @@
...
+       style="fill:#6ea8b3;fill-opacity:1;fill-rule:nonzero;stroke:none"
+       d="m 268.0807,261.8344 c 0,31.6281 25.647,57.2677 57.2842,57.2677 31.6372,0 57.2843,-25.6396 57.2843,-57.2677 0,-31.6281 -25.6471,-57.2677 -57.2843,-57.2677 -31.6372,0 -57.2842,25.6396 -57.2842,57.2677 z" /><path
+       d="m 268.0807,261.83438 c 0,31.62809 25.64703,57.26772 57.28426,57.26772 31.63723,0 57.28425,-25.63963 57.28425,-57.26772 0,-31.6281 -25.64702,-57.26772 -57.28425,-57.26772 -31.63723,0 -57.28426,25.63962 -57.28426,57.26772 z"
+       style="fill:none;stroke:#004957;stroke-width:1.60394;stroke-linecap:butt;stroke-linejoin:round;stroke-miterlimit:10;stroke-dasharray:none;stroke-opacity:1"
+       id="path80"
+       inkscape:connector-curvature="0" /><text
+       y="-263.59802"
+       x="287.62537"
+       transform="scale(1.0001444,-0.99985561)"
+       style="font-variant:normal;font-weight:normal;font-size:13.998px;font-family:Arial;-inkscape-font-specification:ArialMT;writing-mode:lr-tb;fill:#ffffff;fill-opacity:1;fill-rule:nonzero;stroke:none;stroke-width:0.239965"
```
Comment (afedulov) [inline images/statefun-workshop.svg:157]: Forgot one thing. We should try to stick to the same font throughout our material. I believe David settled on Open Sans for the training.

**Comment by sjwiesman** [comment]
I've applied your suggestions and merged into 'solution'
